### ISSUE TITLE
fix: Newton/Annealing/Metropolis step() leak computational graph — missing torch.no_grad()

### ISSUE THREAD

**Issue by kovacoj** [kovacoj1/optimizers#66]
## Bug

`Newton.step()`, `Annealing.step()`, and `Metropolis.step()` lack the `@torch.no_grad()` decorator that PyTorch optimizers conventionally use. The gradient computations inside these methods are done explicitly via `grad()`, but the final `update_weights()` / `_set_params()` calls happen inside a graph-building context.

This means every step leaks the computational graph from the closure evaluation and the parameter update, causing **unbounded memory growth** in a training loop.

For comparison, `LevenbergMarquardt.step()`, `ExtendedKalmanFilter.step()`, and `KalmanFilter.step()` also lack the decorator, but their closure evaluations are wrapped with `torch.enable_grad()`, which at least ensures the graph is built intentionally. Newton, Annealing, and Metropolis don't even have that.

## Expected behavior

PyTorch optimizers (e.g., `torch.optim.SGD.step()`) use `@torch.no_grad()` and explicitly re-enable grad only where needed. The same pattern should apply here:

```python
@torch.no_grad()
def step(self, closure):
    with torch.enable_grad():
        loss = closure()
    # ... gradient computation and update ...
```

This ensures:
1. No graph is retained after `step()` returns.
2. Memory usage is stable across many iterations.
3. The API matches the convention numerical PyTorch users expect.

## Affected files

- `Newton.py:62`
- `Annealing.py:33`
- `Metropolis.py:35`
- `LevenbergMarquardt.py:207`
- `ExtendedKalmanFilter.py:95`
- `KalmanFilter.py:91`

**Comment by kovacoj** [comment]
## Proposed fix

Add `@torch.no_grad()` to all `step()` methods, and use `with torch.enable_grad():` only where the closure or explicit gradient computation needs it:

**Newton:**
```python
@torch.no_grad()
def step(self, closure):
    assert len(self.param_groups) == 1
    params = trainable_params(self.param_groups)
    prototype = params[0]

    with torch.enable_grad():
        grads = grad(closure(), params, create_graph=True, allow_unused=True)
    # ... rest of Hessian computation and update
```

Note: Newton's Hessian computation uses `create_graph=True` which requires grad. The `@torch.no_grad()` on `step` combined with `torch.enable_grad()` for the closure/grad calls is the correct pattern — it ensures the final weight update is in no-grad context while the intermediate gradient computation still works.

**Annealing/Metropolis:**
```python
@torch.no_grad()
def step(self, closure):
    variants = [self.params, self.mutate().clone()]
    Fs = torch.empty(2)
    for idx, params in enumerate(variants):
        self.update_weights(params)
        with torch.enable_grad():
            Fs[idx] = closure().item()
    # ... acceptance and temperature decay
```

**LevenbergMarquardt/ExtendedKalmanFilter/KalmanFilter:**
These already wrap `closure = torch.enable_grad()(closure)`. Just add `@torch.no_grad()` to `step()` and keep the existing `torch.enable_grad()` wrapper. The two work together: `@torch.no_grad()` sets the default context, and `torch.enable_grad()` re-enables grad locally for the closure.

**Genetic:**
```python
@torch.no_grad()
def step(self, closure):
    # ... existing logic
    for idx in range(self.pop_size):
        self.update_weights(self.population[idx])
        with torch.enable_grad():
            loss[idx] = self.directional(closure)
    # ... selection and reproduction
```

Note: `Genetic.directional()` calls `closure().backward()` and `self.helper.step()`, which both need grad enabled. The `torch.enable_grad()` wrapper must cover the entire `directional()` call.

**Verification:** After this change, run a 100-step training loop and verify memory usage is stable (not growing linearly).

**Comment by kovacoj** [comment]
Closing this as not planned after re-checking the implementation.

The concrete reproducible issue here was inconsistent behavior under an outer `torch.no_grad()` context. That is addressed by PR #81 for Newton, which was the optimizer that actually failed this pattern.

The broader memory-leak claim in this issue is not accurate as written:

- Annealing and Metropolis already have `@torch.no_grad` decorators.
- Parameter writes in Newton/LM/EKF/KF go through helpers decorated with `@torch.no_grad()`.
- LM/EKF/KF intentionally re-enable grad for residual/Jacobian construction and do not retain the graph after `step()` returns.

If we later find a concrete retained-graph leak with a reproducer, we should file that as a separate bug with a failing memory or graph-retention test.
